[webkitflaskpy] Add IPCManager object

JonWBedard opened the pull request:
#### ecafc68a10f3675eabed1bf7592ba2e21e1c12fa
<pre>
[webkitflaskpy] Add IPCManager object
<a href="https://bugs.webkit.org/show_bug.cgi?id=253124">https://bugs.webkit.org/show_bug.cgi?id=253124</a>
rdar://106059293

Reviewed by Dewei Zhu.

Implement a generalized version of a Redis queue to pass tasks to worker processes.
These sorts of queues are most often used by web services to handle incoming hooks.

* Tools/Scripts/libraries/webkitflaskpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/__init__.py: Export IPCManager class.
* Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/ipc_manager.py: Added.
(IPCManager.encode): Return encoded dictionary and database key.
(IPCManager.__init__): Construct manager with a list of databases.
(IPCManager.register): Register a callback with a specific name.
(IPCManager.enqueue): Add a task to be run with arguments, number of retries and expiration.
(IPCManager.dequeue): Execute the specified task. If the execution is successful, delete the task.
Otherwise, decrement the retry counter for that task before raising the thrown exception.
(IPCManager.dequeue_all): Execute all tasks currently in the queue.
(IPCManager.tasks): List all tasks in the queue.
* Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/tests/ipc_manager_unittest.py: Added.
(IPCManagerUnittest.increment):
(IPCManagerUnittest.decrement):
(IPCManagerUnittest.test_encoding):
(IPCManagerUnittest.test_enqueue):
(IPCManagerUnittest.test_dequeue):
(IPCManagerUnittest.test_dequeue_all):
(IPCManagerUnittest.test_retry):

Canonical link: <a href="https://commits.webkit.org/261343@main">https://commits.webkit.org/261343@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/547cd5bd3bd20944f75d649848e8132ed98cd3e8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/6/builds/111372 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/20510 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/3067 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/WinCairo-EWS "Waiting in queue, processing has not started yet") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/115326 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/21877 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/11603 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/87/builds/3067 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/19/builds/117136 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/21877 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/104040 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/5/builds/114801 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/21877 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/104040 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/99908 "Build is in progress. Recent messages:") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/13011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/104040 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/13521 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/11603 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/18964 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/15480 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4305 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

dewei-zhu commented on Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/ipc_manager.py:
```diff
@@ -0,0 +1,122 @@
+         encoded = json.dumps([name, args, kwargs, retry, expiration])
+         digest = hashlib.md5()
+         digest.update(string_utils.encode(encoded))
+         digest = digest.hexdigest()
```
I think we can just do `digest = hashlib.md5(string_utils.encode(encoded)).hexdigest()`

dewei-zhu commented on Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/ipc_manager.py:
```diff
@@ -0,0 +1,122 @@
+         if name not in self.callbacks:
+             raise ValueError("'{}' is not a registered callback".format(name))
+         encoded = json.dumps([name, args, kwargs, retry, expiration])
+         digest = hashlib.md5()
```
I guess it's safe enough for md5 as a key rather than a checksum.

JonWBedard replied:
Right, we're not using a hash for security reasons here. We just want a well defined hash function that will be the same on all versions of Python (in case our web and worker processes use different versions of Python, which they shouldn't, that would be nonsense, but I can see some cases where that might be true as we're deploying, for example)

dewei-zhu commented on Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/ipc_manager.py:
```diff
@@ -0,0 +1,122 @@
+         retry = kwargs.pop('retry', self.RETRY)
+         if name not in self.callbacks:
+             raise ValueError("'{}' is not a registered callback".format(name))
+         encoded = json.dumps([name, args, kwargs, retry, expiration])
```
Does this mean the arguments must be json serializable?

dewei-zhu replied:
Also, should we use `json.dumps` with `sort_keys=True`, also worth adding a unit test for this.
```
def foo(**kwargs):
     print(json.dumps(kwargs))
 
foo(a=1, b=2) # {"a": 1, "b": 2}
foo(b=2, a=1) # {"b": 2, "a": 1}
```

Also, what's the expected behavior if I enqueue same request twice or merge the request?

JonWBedard replied:
- Yes, arguments would need to be json serializable. In practice, that's an easy requirement to meet because we use this class when receiving an upload, our arguments are essentially always strings.

- Good call on the `sort_keys=True`, will make that change.

- Expected behavior is that we would end up with a single request. Consider, for example, the results database. If we get two identical uploads, we only need to process those once. Likewise, if we get two of the exact same hook from GitHub at the same time, we only want to process one of them.

dewei-zhu replied:
What if one of those two identical uploads arrived after the first one is dequeued? If the callback throw an exception, we will hit the retry logic until it exhausted retries. Or the callback can handle this gracefully.

Or we just use this nice dedupe as an optimization to avoid same request enqueued twice if it's currently in the queue.

JonWBedard replied:
If two identical uploads arrive after the first one is dequeued, we have 2 basic options:
- The dequeue completed
- The dequeue triggered a retry
In the first case, we'll re-queue the same task and repeat the operation. Note that case 1 covers the task completing successfully, the task being somehow invalid (like the named task doesn't exist on the worker) and the task exhausting its retry.

In the second case, the incoming task will overwrite the retry count, basically reseting it.

There is also the question of what happens if something arrives _during_ the dequeue, which looks very much like the case where two identical tasks are collapsed into one.

dewei-zhu commented on Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/ipc_manager.py:
```diff
@@ -0,0 +1,122 @@
+         digest = digest.hexdigest()
+
+         for database in self.databases:
+             database.set(digest, encoded, ex=expiration)
```
Just to confirm, `database.set` guarantees the atomic operation here?

JonWBedard replied:
Inside Redis, yes. The function call isn't atomic (we do some mutating of the digest to give it a partition key), but that actual insertion into the database is.

dewei-zhu commented on Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/ipc_manager.py:
```diff
@@ -0,0 +1,122 @@
+                 return False
+
+             if retry > 0:
+                 self.databases[index].set(key, json.dumps([name, args, kwargs, retry - 1, expiration]), ex=expiration)
```
Ditto, we may want to `sort_keys=True`. Since we keep repeating  json.dumps in this file and unit test, I guess it would be more reasonable to have a helper function called `encode_task` which produces encoded task. It would also be nice to have `task_key` function, which centralizes the key calculation and can be used in unit test as well.

dewei-zhu commented on Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/tests/ipc_manager_unittest.py:
```diff
@@ -0,0 +1,123 @@
+        manager.enqueue(self.INCREMENT, 2)
+        self.assertEqual(2, len(list(manager.tasks())))
+
+        manager.enqueue(self.DECREMENT, 1)
```
If I do one more `manager.enqueue(self.INCREMENT, 1)` would the queue size stay unchanged since we enqueued same type of request already?

JonWBedard replied:
Correct, will add that in this test too.

dewei-zhu replied:
That's an interesting design. Meaning that counter can be different based on the execution order.
```
manager.enqueue(self.INCREMENT, 1)
manager.enqueue(self.INCREMENT, 1)
manager.dequeue('9ffcb5a7b14e4536e69f6527b1283237')
manager.dequeue('9ffcb5a7b14e4536e69f6527b1283237')
```
We will ends with COUNTER=1

If we do
```
manager.enqueue(self.INCREMENT, 1)
manager.dequeue('9ffcb5a7b14e4536e69f6527b1283237')
manager.enqueue(self.INCREMENT, 1)
manager.dequeue('9ffcb5a7b14e4536e69f6527b1283237')
```
We will ends with COUNTER=2

Is our particular use case fine with this behavior?

JonWBedard replied:
I'm going to enshrine this example in the unit tests.

The short answer to this is "at the moment, and in the immediate future, yes".

The service that uses this now is bug-bridge.webkit.org, and this behavior is acceptable there. Re-doing a task costs compute and network, but is otherwise safe to repeat. If given the option, we'd rather save compute and network.

I plan to use this in commits.webkit.org, where repeating tasks basically means repeating pushes we're already performed, and results.webkit.org, which is designed to be able to re-process uploads.

dewei-zhu commented on Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/tests/ipc_manager_unittest.py:
```diff
@@ -0,0 +1,123 @@
+from webkitflaskpy import IPCManager
+
+
+class IPCManagerUnittest(unittest.TestCase):
```
Would it very complicated to have a unit test that test concurrent situation?

JonWBedard replied:
I don't know that you could write one without bringing up a real instance of Redis. We're leveraging an in-memory approximation of Redis for unit testing, and while real Redis is safe with multiple concurrent connections, the in memory version is not.

dewei-zhu replied:
Got it

dewei-zhu reviewed (approved):
r=me

JonWBedard committed:
[webkitflaskpy] Add IPCManager object
https://bugs.webkit.org/show_bug.cgi?id=253124
rdar://106059293

Reviewed by Dewei Zhu.

Implement a generalized version of a Redis queue to pass tasks to worker processes.
These sorts of queues are most often used by web services to handle incoming hooks.

* Tools/Scripts/libraries/webkitflaskpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/__init__.py: Export IPCManager class.
* Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/ipc_manager.py: Added.
(IPCManager.encode): Return encoded dictionary and database key.
(IPCManager.__init__): Construct manager with a list of databases.
(IPCManager.register): Register a callback with a specific name.
(IPCManager.enqueue): Add a task to be run with arguments, number of retries and expiration.
(IPCManager.dequeue): Execute the specified task. If the execution is successful, delete the task.
Otherwise, decrement the retry counter for that task before raising the thrown exception.
(IPCManager.dequeue_all): Execute all tasks currently in the queue.
(IPCManager.tasks): List all tasks in the queue.
* Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/tests/ipc_manager_unittest.py: Added.
(IPCManagerUnittest.increment):
(IPCManagerUnittest.decrement):
(IPCManagerUnittest.test_encoding):
(IPCManagerUnittest.test_enqueue):
(IPCManagerUnittest.test_dequeue):
(IPCManagerUnittest.test_dequeue_all):
(IPCManagerUnittest.test_retry):

Canonical link: https://commits.webkit.org/261343@main
